<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta http-equiv="X-UA-Compatible" content="ie=edge">
  <title>取消请求</title>
</head>
<body>
  <button onclick="getProducts1()">获取商品列表1</button><br>
  <button onclick="getProducts2()">获取商品列表2</button><br>
  <button onclick="cancelReq()">取消请求</button><br>

  <script src="https://cdn.bootcss.com/axios/0.19.0/axios.js"></script>
  <script>

    // 场景：反复发送请求的情况下，需要取消前面的，而发送最新的请求
    // 所以需要在准备发请求前, 取消未完成的请求
    // 但是单单的在请求前做判断的话，还是不够的
    // 因为如果多次点击取消，后面的取消不会响应

    // 原因：第一次发送请求后继续点击
    // 这个时候进入了取消判断，发现cancel是函数，因此执行了取消请求
    // 进入到了error处理中，把cancel置为了 null 
    // (这个是关键点，cancel是回调函数，因此放到异步队列，先执行下面的代码，而cancelToken 里面的执行器是同步代码，因此先将回调函数赋值了，覆盖了)
    // (然后执行error，在里面把cancel置为null，而第二次的cancel也被重置了)
    // 第三次点击，发现第二次的cancel是null，因此没有取消任务

    // 所以就会出现隔一次再取消的情况

    // 所以在取消中还需要再判断一下
    // 判断当前error调用的是 取消 还是 错误
    


    let cancel;  // 用于保存取消请求的函数
    function getProducts1() {
      // 在准备发请求前, 取消未完成的请求
      if (typeof cancel==='function') {
        cancel('取消请求');
      }

      axios({
        url: 'http://localhost:4000/products1',
        cancelToken: new axios.CancelToken((c) => { // c是用于取消当前请求的函数
          // 保存取消函数, 用于之后可能需要取消当前请求
          cancel = c;
        })
      }).then(
        response => {
          cancel = null;
          console.log('请求1成功了', response.data);
        },
        error => {
          if (axios.isCancel(error)) {
            // cancel = null;
            console.log('请求1取消的错误', error.message);
          } else { // 请求出错了
            cancel = null;
            console.log('请求1失败了', error.message);
          }
          
        }
      )
    }

    function getProducts2() {
      // 在准备发请求前, 取消未完成的请求
      if (typeof cancel==='function') {
        cancel('取消请求');
      }

      axios({
        url: 'http://localhost:4000/products2',
        cancelToken: new axios.CancelToken((c) => { // c是用于取消当前请求的函数
          // 保存取消函数, 用于之后可能需要取消当前请求
          cancel = c;
        })
      }).then(
        response => {
          cancel = null;
          console.log('请求2成功了', response.data);
        },
        error => {
          if (axios.isCancel(error)) {
            // cancel = null;
            console.log('请求2取消的错误', error.message);
          } else { // 请求出错了
          }
        }
      )
    }

    function cancelReq() {
      // alert('取消请求');
      // 执行取消请求的函数
      if (typeof cancel === 'function') {
        cancel('强制取消请求');
      } else {
        console.log('没有可取消的请求');
      }
    }
  </script>
</body>
</html>